src/Microsoft.Kusto.ServiceLayer/DataSource/Monitor/MonitorDataSource.cs

```diff
@@ -0,0 +1,198 @@
+            {
+                var tableGroup = metadataTableGroups[workspaceTableGroup];
+
+                var tableGroupNodeInfo =
+                    MetadataFactory.CreateDataSourceObjectMetadata(DataSourceMetadataType.Folder, tableGroup.Name, $"{workspace.Id}.{tableGroup.Name}");
+
+                _nodes.SafeAdd($"{workspace.Id}", tableGroupNodeInfo);
+                
+                SetupTables(tableGroupNodeInfo);
+            }
+        }
+        
+        private void SetupTables(DataSourceObjectMetadata tableGroupNodeInfo)
+        {
+            var tables = GetNonEmptyTableNames();
+            var metadataTables = _metadata.Tables.ToDictionary(x => x.Name);
+            
+            foreach (string tableName in tables)
+            {
+                var table = metadataTables[tableName];
+
+                var tableNodeInfo = MetadataFactory.CreateDataSourceObjectMetadata(DataSourceMetadataType.Table, table.Name,
+                    $"{tableGroupNodeInfo.Urn}.{table.Name}");
+
+                _nodes.SafeAdd(tableGroupNodeInfo.Urn, tableNodeInfo);
+
+                SetupColumns(table, tableNodeInfo);
+            }
+        }
+        
+        private IEnumerable<string> GetNonEmptyTableNames()
+        {
+            string query = "union * | summarize count() by Type";
+            var results = _monitorClient.Query(query);
+            return results.Tables[0].Rows.Select(x => x[0]).OrderBy(x => x);
+        }
+        
+        private void SetupColumns(TablesModel table, DataSourceObjectMetadata tableNodeInfo)
+        {
+            foreach (var column in table.Columns)
+            {
+                var columnNodeInfo = MetadataFactory.CreateDataSourceObjectMetadata(DataSourceMetadataType.Column, column.Name,
+                    $"{tableNodeInfo.Urn}.{column.Name}");
+
+                _nodes.SafeAdd(tableNodeInfo.Urn, columnNodeInfo);
+            }
+        }
+
+        public override async Task<IDataReader> ExecuteQueryAsync(string query, CancellationToken cancellationToken, string databaseName = null)
+        {
+            var results =  await _monitorClient.QueryAsync(query, cancellationToken);
+            return results.ToDataReader();
+        }
+
+        public override Task<IEnumerable<T>> ExecuteControlCommandAsync<T>(string command, bool throwOnError, CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override DiagnosticsInfo GetDiagnostics(DataSourceObjectMetadata parentMetadata)
+        {
+            return new DiagnosticsInfo();
+        }
+
+        public override IEnumerable<DataSourceObjectMetadata> GetChildObjects(DataSourceObjectMetadata parentMetadata, bool includeSizeDetails = false)
+        {
+            // columns are always leaf nodes
+            if (parentMetadata.MetadataType == DataSourceMetadataType.Column)
+            {
+                return Enumerable.Empty<DataSourceObjectMetadata>();
```

Comment:
Kusto Data Source returns an exception if the type is not expected. Should we do that instead of returning an empty list.

Reply:
This just mimics the behavior of the KustoDataSource since I implemented this datasource differently.